### PR TITLE
use all custom reporter event listeners

### PR DESCRIPTION
**Pull request by gronke**
Creates Allure reports on runtime by using all [custom reporter events](http://webdriver.io/guide/testrunner/customreporter.html#Events).

#### Features
- catches errors in hooks with a virtual added test case that only shows up on failure
- group tests and sub-suites within root suite
- nested suites will concatenate with parent names up (except the root suite)

### CODE REVIEW

**Comment by christian-bromann** [comment]
@just-boris any feedback?

File: lib/reporter.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import events from 'events'
 import Allure from 'allure-js-commons'
 import Runtime from 'allure-js-commons/runtime'
+import Suite from 'allure-js-commons/beans/suite'
```
Comment (just-boris) [inline lib/reporter.js]: Unused import?

File: lib/reporter.js [hunk 2]
```diff
@@ -25,190 +30,133 @@ class AllureReporter extends events.EventEmitter {
...
+        })
 
-    formatTime (time) {
-        return `${this.padTime(time.getHours())}:${this.padTime(time.getMinutes())}:${this.padTime(time.getSeconds())}.${this.padTime(time.getMilliseconds(), 3)}`
-    }
+        this.on('test:start', (test) => {
+            var title = this.getTitle(test)
+            this.pushParent(title)
+            this.print('test:start >', title)
+            this.allure.startCase(title, this.getTime())
```
Comment (just-boris) [inline lib/reporter.js]: If you do not pass the time, Allure uses `Date.now()` by default. You can get rid of `getTime()` method at all, when you are doing real-time reporting.

**Comment by just-boris** [comment]
This approach looks better from the point of interaction with Allure. For example, now you can build a report with some suites from very long test run while other suites are running.

But the lack of screenshots makes me sad. I think it can be implemented with `runner:aftercommand`, but I didn't find how does it work.

**Comment by gronke** [comment]
[BaseReporter.js#L82-L84](https://github.com/webdriverio/webdriverio/blob/master/lib/utils/BaseReporter.js#L82-L84) looks promising. We should flag this as "do-not-merge" until we have support for screenshots and the other few unhandled events left. Work in progress.

**Comment by christian-bromann** [comment]
no big deal

**Comment by just-boris** [comment]
As I understand, the `runner:screenshot` is fired only when we are taking a screenshot on failed test. Screenshots made with `browser.saveScreenshot()` still missing.

File: lib/reporter.js [hunk 3]
```diff
@@ -110,6 +110,17 @@ class AllureReporter extends events.EventEmitter {
             this.currentHooks.pop();
         })
 
+        this.on('runner:screenshot', (screenshot) => {
+
+            var filename = this.getScreenshotFilename(screenshot),
+                data = new Buffer(screenshot.data, 'base64')
+
+            this.print('screenshot', filename)
+            this.allure.addAttachment(filename, data)
```
Comment (just-boris) [inline lib/reporter.js]: FYI, filename is not an actual filename. It is just a string to display this screenshot in attachment list.

Comment (gronke) [inline lib/reporter.js]: Afaik at least the extension should match the mime type. It feels like a filename of the attached asset.

Do you have a suggestion for better auto-naming? The timestamp maybe? I'm not really sure what suites the best.

Comment (just-boris) [inline lib/reporter.js]: We are using [file-type](https://github.com/sindresorhus/file-type) for auto-detect file extention if `type` argument is empty.

>Do you have a suggestion for better auto-naming? 
The best option will be `filename` value of the argument from [saveScreenshot](http://webdriver.io/api/utility/saveScreenshot.html) command

**Comment by gronke** [comment]
@just-boris it's exactly the opposite :smile_cat:. Taking screenshots manually via `browser.saveScreenshot()` or `browser.saveScreenshot('example.png')` passes the files to Allure. I'm still having problems getting screenshots taken on failure.  After getting that working I expect screenshots during failed test-cases to be propagated, but an issue when a hook fails and a screenshot is taken.

To support instances running in parallel it was necessary to queue allure commands per suite/instance and flush the queue shortly before writing the report. That's because Allure does only like to work at one case and suite at a time.

**Comment by just-boris** [comment]
https://github.com/webdriverio/webdriverio/blob/master/lib/webdriverio.js#L221: the same event is fired as well in saveScreenshot command. I don't know why it is not caught.

Note about parallel: the best thing in Allure architecture is that report consists from serval XMLs, one for each suite. So, for parallel runs, the best solution will be a separate Allure instance for each thread. Each instance writes a separate report and it will be concatenated into single page during `allure generate`.

**Comment by gronke** [comment]
@just-boris indeed, running multiple instances of Allure in parallel would work. I'm running everything in a single instance instance to later get feature/story-label support. Looks we can easily switch to multiple instances when it becomes necessary. That could be the case when two tests end at the same time, before Allure finished writing it's files.

https://github.com/webdriverio/webdriverio/issues/1249 is necessary, because at the time the `runner:screenshow` was fired, the test already ended with a failure. With the new order we can first attach the screenshot and then end the test case.

**Comment by just-boris** [comment]
@gronke see a related issue about stories: https://github.com/webdriverio/wdio-allure-reporter/issues/11

There is my concept that we can provide stories name as a part of test title.

In another case, we can create a custom command `browser.setStory()`, that fires event for Allure reporter. Maybe we should keep it separately, not in `browser` object, but in special global `allure` like other our test plugins doing

**Comment by christian-bromann** [comment]
@gronke what's the state of this PR? Seems that there will be quite some changes to be introduced. Two things I've see while scrolling through:

- could keep file names lowercase?
- could you add small test coverage where it makes sense?

**Comment by gronke** [comment]
@christian-bromann work is still in progress. Since the initial approach of binding to the BaseReporter events did not work when errors occurred in task-related hooks or when taking screenshots. Currently I'm running the reporter on existing test-suites to determine edge cases and other sources of errors. Based on that my plan was to create unit tests that properly define the desired behavior we can then discuss.

**Comment by sullenor** [comment]
Hi, how is going? I wonder if you need any help here?

**Comment by gronke** [comment]
Was a bit short of time the last few days, but in the meanwhile this is used in two projects of mine and it works quite well so far. There is still (at least) one condition where `stop=""` in a result let's the allure generator fail. Will have time to investigate details from the build logs to this issue.

Before this can be merged unit tests are mandatory. Think it makes the most sense to pick up the discussion how this reporter should apply the test suite structure to Allure. For example a `"before all" hook` is only reported to Allure if it fails, because otherwise the results would be flooded. There are quite a few things that are opinions and need to be reviewed, but all that is the easiest with some unit tests to talk about.

@sullenor thank you for asking! Let's look into this after having a set of unit tests that I can add by end of this weekend.

**Comment by gronke** [comment]
@sullenor as promised here's the PR for unit tests #12. Waiting for feedback on that setup now to then apply more test cases that nail down the behavior of this branch.

**Comment by christian-bromann** [comment]
@gronke thanks for all the work! Excited about seeing this getting published

**Comment by christian-bromann** [comment]
@gronke @sullenor any objections merging and releasing the current two PRs? It seems that the changes are stable according to @gronke comment.

**Comment by gronke** [comment]
@christian-bromann :blush: thanks for the flowers.

Now that the @just-boris helped refactoring the integration test suite I'd like to rebase against that branch and describe all features of this reporter in tests. Even if that takes a bit more time, I think a test utility like this should proove its stability first. There will be many general tests of screenshots, hooks and attachments, but also tests for opinionated behavior that will probably need some discussion.

What I'm not sure about is
- Top level suites become an Allure suite. All further nested suite and test titles are concatenated to become an Allure case.
- Hooks are only listed if broken. (can hooks also fail or only break?)

**Comment by christian-bromann** [comment]
> can hooks also fail or only break?

Yeah I think, if something throws an error inside

Ok cool, looking forward to have all the new stuff.

**Comment by sullenor** [comment]
@christian-bromann @gronke I tried to reproduce my issues with fresh version of webdriverIO and wdio-allure-reporter and seems to me that everything works as it should be. So, you did a great work here, guys!

**Comment by gronke** [comment]
This branch still does not pass all of the integration tests and needs some refactoring. In a few more days I will be able to push an update targeting the failing tests.

Everybody is welcome to pair program on the refactoring part. I can offer Mumble and tmux.

**Comment by anarwal** [comment]
@gronke  Is there anything I cant help you with in this?

**Comment by gronke** [comment]
@anarwal thanks so much for asking. Right now I'm on a trip in Sweden ✈️ , but will be back in about a week and continue working on this branch asap. In case you already have expectations/ideas for more tests, please feel free to add some in a PR.

**Comment by gronke** [comment]
#17 :yellow_heart: 

@just-boris will merge the changes into my ongoing work. Will be able to continue this week and plan as next step simplify the stack structure because it feels too complex for the problem.

**Comment by just-boris** [comment]
@gronke, I think current state of #17 can be safely merged and it should work as before.

Then we can send new updates in further PRs

**Comment by anarwal** [comment]
@just-boris @christian-bromann would this merge resolve issue of hooks showing as broken in allure report?
